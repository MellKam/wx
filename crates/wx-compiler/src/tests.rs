use super::*;
use indoc::indoc;
use insta::{assert_binary_snapshot, assert_yaml_snapshot};
use string_interner::{StringInterner, backend::StringBackend};

#[allow(unused)]
struct TestCase {
    files: files::Files,
    interner: StringInterner<StringBackend>,
    ast: ast::parser::ParserResult,
    hir: hir::builder::BuilderResult,
    mir: mir::MIR,
    module: wasm::Module,
    bytecode: Vec<u8>,
}

impl TestCase {
    fn assert_snapshot(name: &str, source: &str) {
        let mut interner = StringInterner::new();
        let mut files = files::Files::new();
        let file_id = files
            .add(name.to_string() + ".wx", source.to_string())
            .unwrap();

        let ast =
            ast::parser::Parser::parse(file_id, &files.get(file_id).unwrap().source, &mut interner);
        assert_yaml_snapshot!(name.to_string() + "_ast", ast);

        if ast.diagnostics.len() > 0 {
            return;
        }

        let hir = hir::Builder::build(&ast.ast, &mut interner);
        assert_yaml_snapshot!(name.to_string() + "_hir", hir);

        let mir = mir::Builder::build(&hir.hir);
        assert_yaml_snapshot!(name.to_string() + "_mir", mir);

        let module = wasm::Builder::build(&mir).unwrap();
        assert_yaml_snapshot!(name.to_string() + "_wasm", module);

        let bytecode = wasm::Encoder::encode(&module, &interner);
        assert_binary_snapshot!((name.to_string() + ".wasm").as_ref(), bytecode.clone());
    }
}

#[test]
fn factorial() {
    TestCase::assert_snapshot(
        "factorial",
        indoc! {"
        export func factorial(n: i32): i32 {
            if n <= 1 { return 1 };
            n * factorial(n - 1)
        }
    "},
    );
}

#[test]
fn pow() {
    TestCase::assert_snapshot(
        "pow",
        indoc! {"
        export func pow(base: i32, mut exp: i32): i32 {
            if exp < 0 {
                return 0; 
            };
            
            local mut result: i32 = 1;
            loop {
                if exp == 0 {
                break result; 
                };
                
                result *= base;
                exp -= 1;
            }
        }
    "},
    );
}

#[test]
fn func_pointers() {
    TestCase::assert_snapshot(
        "func_pointers",
        indoc! {"
        func add(a: i32, b: i32): i32 { a + b }
        func sub(a: i32, b: i32): i32 { a - b }

        func apply(
            binop: func(i32, i32) -> i32, 
            a: i32, 
            b: i32,
        ): i32 {
            binop(a, b)
        }

        export func main(): i32 {
            local a = apply(add, 5, 10);
            local b = apply(sub, 10, 5);
            local c: i32 = 5;

            a + b
        }
    "},
    );
}
