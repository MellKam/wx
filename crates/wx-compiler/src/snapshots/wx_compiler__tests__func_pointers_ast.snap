---
source: crates/wx-compiler/src/tests.rs
expression: ast
---
ast:
  file_id: 0
  items:
    - kind:
        FunctionDefinition:
          signature:
            name:
              symbol: 1
              span:
                start: 5
                end: 8
            params:
              open:
                start: 8
                end: 9
              inner:
                - inner:
                    mutable: ~
                    name:
                      symbol: 2
                      span:
                        start: 9
                        end: 10
                    annotation:
                      separator:
                        start: 10
                        end: 11
                      ty:
                        kind:
                          Identifier:
                            symbol: 3
                        span:
                          start: 12
                          end: 15
                  span:
                    start: 9
                    end: 15
                  separator:
                    start: 15
                    end: 16
                - inner:
                    mutable: ~
                    name:
                      symbol: 4
                      span:
                        start: 17
                        end: 18
                    annotation:
                      separator:
                        start: 18
                        end: 19
                      ty:
                        kind:
                          Identifier:
                            symbol: 3
                        span:
                          start: 20
                          end: 23
                  span:
                    start: 17
                    end: 23
                  separator: ~
              close:
                start: 23
                end: 24
            result:
              separator:
                start: 24
                end: 25
              ty:
                kind:
                  Identifier:
                    symbol: 3
                span:
                  start: 26
                  end: 29
          block:
            kind:
              Block:
                open:
                  start: 30
                  end: 31
                inner:
                  - inner:
                      Expression:
                        expr:
                          kind:
                            Binary:
                              left:
                                kind:
                                  Identifier:
                                    symbol: 2
                                span:
                                  start: 32
                                  end: 33
                              operator:
                                kind: Add
                                span:
                                  start: 34
                                  end: 35
                              right:
                                kind:
                                  Identifier:
                                    symbol: 4
                                span:
                                  start: 36
                                  end: 37
                          span:
                            start: 32
                            end: 37
                    span:
                      start: 32
                      end: 37
                    separator: ~
                close:
                  start: 38
                  end: 39
            span:
              start: 30
              end: 39
      span:
        start: 0
        end: 39
    - kind:
        FunctionDefinition:
          signature:
            name:
              symbol: 5
              span:
                start: 45
                end: 48
            params:
              open:
                start: 48
                end: 49
              inner:
                - inner:
                    mutable: ~
                    name:
                      symbol: 2
                      span:
                        start: 49
                        end: 50
                    annotation:
                      separator:
                        start: 50
                        end: 51
                      ty:
                        kind:
                          Identifier:
                            symbol: 3
                        span:
                          start: 52
                          end: 55
                  span:
                    start: 49
                    end: 55
                  separator:
                    start: 55
                    end: 56
                - inner:
                    mutable: ~
                    name:
                      symbol: 4
                      span:
                        start: 57
                        end: 58
                    annotation:
                      separator:
                        start: 58
                        end: 59
                      ty:
                        kind:
                          Identifier:
                            symbol: 3
                        span:
                          start: 60
                          end: 63
                  span:
                    start: 57
                    end: 63
                  separator: ~
              close:
                start: 63
                end: 64
            result:
              separator:
                start: 64
                end: 65
              ty:
                kind:
                  Identifier:
                    symbol: 3
                span:
                  start: 66
                  end: 69
          block:
            kind:
              Block:
                open:
                  start: 70
                  end: 71
                inner:
                  - inner:
                      Expression:
                        expr:
                          kind:
                            Binary:
                              left:
                                kind:
                                  Identifier:
                                    symbol: 2
                                span:
                                  start: 72
                                  end: 73
                              operator:
                                kind: Sub
                                span:
                                  start: 74
                                  end: 75
                              right:
                                kind:
                                  Identifier:
                                    symbol: 4
                                span:
                                  start: 76
                                  end: 77
                          span:
                            start: 72
                            end: 77
                    span:
                      start: 72
                      end: 77
                    separator: ~
                close:
                  start: 78
                  end: 79
            span:
              start: 70
              end: 79
      span:
        start: 40
        end: 79
    - kind:
        FunctionDefinition:
          signature:
            name:
              symbol: 6
              span:
                start: 86
                end: 91
            params:
              open:
                start: 91
                end: 92
              inner:
                - inner:
                    mutable: ~
                    name:
                      symbol: 7
                      span:
                        start: 97
                        end: 102
                    annotation:
                      separator:
                        start: 102
                        end: 103
                      ty:
                        kind:
                          Function:
                            params:
                              open:
                                start: 108
                                end: 109
                              inner:
                                - inner:
                                    kind:
                                      Identifier:
                                        symbol: 3
                                    span:
                                      start: 109
                                      end: 112
                                  span:
                                    start: 109
                                    end: 112
                                  separator:
                                    start: 112
                                    end: 113
                                - inner:
                                    kind:
                                      Identifier:
                                        symbol: 3
                                    span:
                                      start: 114
                                      end: 117
                                  span:
                                    start: 114
                                    end: 117
                                  separator: ~
                              close:
                                start: 117
                                end: 118
                            result:
                              separator:
                                start: 119
                                end: 121
                              ty:
                                kind:
                                  Identifier:
                                    symbol: 3
                                span:
                                  start: 122
                                  end: 125
                        span:
                          start: 104
                          end: 108
                  span:
                    start: 97
                    end: 108
                  separator:
                    start: 125
                    end: 126
                - inner:
                    mutable: ~
                    name:
                      symbol: 2
                      span:
                        start: 132
                        end: 133
                    annotation:
                      separator:
                        start: 133
                        end: 134
                      ty:
                        kind:
                          Identifier:
                            symbol: 3
                        span:
                          start: 135
                          end: 138
                  span:
                    start: 132
                    end: 138
                  separator:
                    start: 138
                    end: 139
                - inner:
                    mutable: ~
                    name:
                      symbol: 4
                      span:
                        start: 145
                        end: 146
                    annotation:
                      separator:
                        start: 146
                        end: 147
                      ty:
                        kind:
                          Identifier:
                            symbol: 3
                        span:
                          start: 148
                          end: 151
                  span:
                    start: 145
                    end: 151
                  separator:
                    start: 151
                    end: 152
              close:
                start: 153
                end: 154
            result:
              separator:
                start: 154
                end: 155
              ty:
                kind:
                  Identifier:
                    symbol: 3
                span:
                  start: 156
                  end: 159
          block:
            kind:
              Block:
                open:
                  start: 160
                  end: 161
                inner:
                  - inner:
                      Expression:
                        expr:
                          kind:
                            Call:
                              callee:
                                kind:
                                  Identifier:
                                    symbol: 7
                                span:
                                  start: 166
                                  end: 171
                              arguments:
                                open:
                                  start: 171
                                  end: 172
                                inner:
                                  - inner:
                                      kind:
                                        Identifier:
                                          symbol: 2
                                      span:
                                        start: 172
                                        end: 173
                                    span:
                                      start: 172
                                      end: 173
                                    separator:
                                      start: 173
                                      end: 174
                                  - inner:
                                      kind:
                                        Identifier:
                                          symbol: 4
                                      span:
                                        start: 175
                                        end: 176
                                    span:
                                      start: 175
                                      end: 176
                                    separator: ~
                                close:
                                  start: 176
                                  end: 177
                          span:
                            start: 166
                            end: 177
                    span:
                      start: 166
                      end: 177
                    separator: ~
                close:
                  start: 178
                  end: 179
            span:
              start: 160
              end: 179
      span:
        start: 81
        end: 179
    - kind:
        ExportModifier:
          export:
            start: 181
            end: 187
          item:
            kind:
              FunctionDefinition:
                signature:
                  name:
                    symbol: 8
                    span:
                      start: 193
                      end: 197
                  params:
                    open:
                      start: 197
                      end: 198
                    inner: []
                    close:
                      start: 198
                      end: 199
                  result:
                    separator:
                      start: 199
                      end: 200
                    ty:
                      kind:
                        Identifier:
                          symbol: 3
                      span:
                        start: 201
                        end: 204
                block:
                  kind:
                    Block:
                      open:
                        start: 205
                        end: 206
                      inner:
                        - inner:
                            LocalDefinition:
                              mutable: ~
                              name:
                                symbol: 2
                                span:
                                  start: 217
                                  end: 218
                              annotation: ~
                              eq:
                                start: 219
                                end: 220
                              value:
                                kind:
                                  Call:
                                    callee:
                                      kind:
                                        Identifier:
                                          symbol: 6
                                      span:
                                        start: 221
                                        end: 226
                                    arguments:
                                      open:
                                        start: 226
                                        end: 227
                                      inner:
                                        - inner:
                                            kind:
                                              Identifier:
                                                symbol: 1
                                            span:
                                              start: 227
                                              end: 230
                                          span:
                                            start: 227
                                            end: 230
                                          separator:
                                            start: 230
                                            end: 231
                                        - inner:
                                            kind:
                                              Int:
                                                value: 5
                                            span:
                                              start: 232
                                              end: 233
                                          span:
                                            start: 232
                                            end: 233
                                          separator:
                                            start: 233
                                            end: 234
                                        - inner:
                                            kind:
                                              Int:
                                                value: 10
                                            span:
                                              start: 235
                                              end: 237
                                          span:
                                            start: 235
                                            end: 237
                                          separator: ~
                                      close:
                                        start: 237
                                        end: 238
                                span:
                                  start: 221
                                  end: 238
                          span:
                            start: 211
                            end: 238
                          separator:
                            start: 238
                            end: 239
                        - inner:
                            LocalDefinition:
                              mutable: ~
                              name:
                                symbol: 4
                                span:
                                  start: 250
                                  end: 251
                              annotation: ~
                              eq:
                                start: 252
                                end: 253
                              value:
                                kind:
                                  Call:
                                    callee:
                                      kind:
                                        Identifier:
                                          symbol: 6
                                      span:
                                        start: 254
                                        end: 259
                                    arguments:
                                      open:
                                        start: 259
                                        end: 260
                                      inner:
                                        - inner:
                                            kind:
                                              Identifier:
                                                symbol: 5
                                            span:
                                              start: 260
                                              end: 263
                                          span:
                                            start: 260
                                            end: 263
                                          separator:
                                            start: 263
                                            end: 264
                                        - inner:
                                            kind:
                                              Int:
                                                value: 10
                                            span:
                                              start: 265
                                              end: 267
                                          span:
                                            start: 265
                                            end: 267
                                          separator:
                                            start: 267
                                            end: 268
                                        - inner:
                                            kind:
                                              Int:
                                                value: 5
                                            span:
                                              start: 269
                                              end: 270
                                          span:
                                            start: 269
                                            end: 270
                                          separator: ~
                                      close:
                                        start: 270
                                        end: 271
                                span:
                                  start: 254
                                  end: 271
                          span:
                            start: 244
                            end: 271
                          separator:
                            start: 271
                            end: 272
                        - inner:
                            LocalDefinition:
                              mutable: ~
                              name:
                                symbol: 9
                                span:
                                  start: 283
                                  end: 284
                              annotation:
                                separator:
                                  start: 284
                                  end: 285
                                ty:
                                  kind:
                                    Identifier:
                                      symbol: 3
                                  span:
                                    start: 286
                                    end: 289
                              eq:
                                start: 290
                                end: 291
                              value:
                                kind:
                                  Int:
                                    value: 5
                                span:
                                  start: 292
                                  end: 293
                          span:
                            start: 277
                            end: 293
                          separator:
                            start: 293
                            end: 294
                        - inner:
                            Expression:
                              expr:
                                kind:
                                  Binary:
                                    left:
                                      kind:
                                        Identifier:
                                          symbol: 2
                                      span:
                                        start: 300
                                        end: 301
                                    operator:
                                      kind: Add
                                      span:
                                        start: 302
                                        end: 303
                                    right:
                                      kind:
                                        Identifier:
                                          symbol: 4
                                      span:
                                        start: 304
                                        end: 305
                                span:
                                  start: 300
                                  end: 305
                          span:
                            start: 300
                            end: 305
                          separator: ~
                      close:
                        start: 306
                        end: 307
                  span:
                    start: 205
                    end: 307
            span:
              start: 188
              end: 307
      span:
        start: 181
        end: 307
diagnostics: []
